HA LLEGADO EL CAPITALISMO AL PROGRAMA DE CONVIVENCIA ESPACIAL

Pj6595 opened the pull request:
**QUEDA CREADO EL MODO CAPITALISMO**

##OVERVIEW DEL MODO CAPITALISMO

El escenario cuenta con roombas que se van moviendo (de manera similar al router) y van spawneando monedas cada cierto tiempo, que se va reduciendo cuanto menos tiempo queda para terminar la ronda. Al hacer daño a un jugador soltará unas cuantas monedas dependiendo del daño que se le haya infligido. Se acaba cuando se acaba el tiempo y gana el que más monedas tenga. En caso de empate se acaba cuando haya un jugador que tenga más monedas que los demás.

### Cambios importantes
->Los players ahora se crean en GameMode
->Los players se crean con la factoría PlayerFactory
->Ahora existe la clase coin (lleva el valor de cada coin y la activa/desactiva)
->Clase Wallet: lanza monedas cuando te hacen daño y lleva la cuenta de tus monedas
->MODO DE JUEGO CAPITALISMO: 
->Añadidas las pools de objetos (se utilizan para las coins, aunque su tamaño no puede determinarse con una constante de json, cosas de plantillas de C++ I guess)
->La capa de colisión Weapon ahora se llama PickableObject

### Otros cambios
->Entity tiene una constructora por defecto (sin entityManager) y EntityManager ahora tiene un método para añadir entidades existentes (utilizado para incluir las entidades de pools)
->Fixes a los GameModes Pelea por el Mando y Stocks (thanks for nothing MMur)

mmur07 committed:
empuje inicial

title

Pj6595 committed:
Creada la PlayerFactory

destruidas nuestras ganas de vivir

mmur07 committed:
empezado wallet

stonks

mmur07 committed:
el jugador puede recoger monedas pero para quitarlas del entity manager es un horror

zoinks

Pj6595 committed:
Implementadas las pools de coins

Por fin hay pools para las coins.
La peculiaridad que tienen es que todos los elementos (en este caos entities) se inicializan con el pool y nosotros les metemos los datos depués, por lo que en Entity hemos tenido que hacer una constructora vacía y en EntityManager un método addExistingentity() para añadir una entidad que ya exista (en este caso las creadas por el pool). Todo esto parece que funciona (de momento), aunque al meterle el size de la pool en el template no admite el macro CONST para cargar una constante de json. De momento es 20 (números mágicos, jeje).

Also, ahora en el PostSolve de CollisionHandler se itera con formas modernas.

mmur07 committed:
el jugador puede recoger monedas y no explota todo

s0nico el erizo

mmur07 committed:
terminado el modo capitalismo. Queda implementarlo con el tilemap

also arreglado el StocksGameMode para que no ganara el jugador que se quedaba en el segundo puesto

Pj6595 committed:
Merge branch 'master' into 171293211-ModoCapitalismo

Pj6595 committed:
Cambios asociados al merge y fixes de última hora

EstebenR commented on src/SourceFiles/EntityManager.cpp:
```diff
@@ -24,10 +24,15 @@ void EntityManager::handleInput() {
+void EntityManager::addExistingEntity(Entity* e) {
+	e->setEntityManager(this);
+	std::unique_ptr<Entity> uPtr(e);
+	entities_.emplace_back(std::move(uPtr));
```
deberia ser un emplace? El emplace crea un objeto, si estamos moviendo el uPtr no deberia ser push?

Pj6595 replied:
Es la pool de Samir y yo no me atrevo a llevarle la contraria a Dios

EstebenR commented on src/SourceFiles/ObjectFactory.h:
```diff
@@ -2,17 +2,25 @@
-class Weapon;
-
-class WeaponFactory
+#include "Transform.h"
+#include "Texture.h"
+#include "Viewer.h"
+#include "Health.h"
+#include "HealthViewer.h"
+#include "InputHandler.h"
+#include "Weapon.h"
+#include "Hands.h"
+#include "Coin.h"
```
Por que está todo esto en el .h y no en el cpp?

mmur07 replied:
Qué problema hay con que esté en el .h? No lo hacemos siempre así excepto cuando pueden salir problemas de inclusividad circular?

EstebenR replied:
Si no hay necesidad lo veo mejor en el cpp, siempre nos dicen que a menos que sea necesario mejor dejarlo en el cpp

EstebenR commented on src/SourceFiles/PCE_Project.vcxproj.filters:
```diff
@@ -174,6 +177,30 @@
+    <ClInclude Include="CoinsPool.h">
+      <Filter>Header Files</Filter>
+    </ClInclude>
+    <ClInclude Include="SRandBasedGenerator.h">
```
No hace falta tener un SRandBasedGenerator, con hacer rand() vale, el srand se hace en el init de SDL_Game

EstebenR commented on src/SourceFiles/SDL_Game.cpp:
```diff
@@ -46,6 +46,8 @@ SDL_Game::~SDL_Game() {
 }
 
 void SDL_Game::initializeResources() {
+	rng_ = new SRandBasedGenerator();
+	rng_->init();
```
mismo del sRandBasedGenerator, sobra

EstebenR commented on src/SourceFiles/StocksGameMode.cpp:
```diff
@@ -12,9 +14,13 @@ StocksGameMode::~StocksGameMode()
-	for(int i = 0;i<players_.size();i++){
+	for (int i = 0; i < nPlayers_; i++) {
+		players_.push_back(PlayerFactory::createPlayerWithHealth(game->getEntityManager(), game->getPhysicsWorld(), i,
+			Resources::Tinky, tilemap_->getPlayerSpawnPoint(i).x, tilemap_->getPlayerSpawnPoint(i).y, 3));
```
Ya tenemos sprite del jugador, no tiene por que ser Tinky

EstebenR commented on src/SourceFiles/Viewer.cpp:
```diff
@@ -66,6 +66,7 @@ void Viewer::init() {
 
 void Viewer::draw() const {
 	if (drawable_) {
+		SDL_Rect r = collider_->getRectRender();
```
¿Para que se usa esto?

Pj6595 replied:
Una línea de prueba que me olvidé de borrar, no hace nada

EstebenR reviewed (changes requested):
Muchas cosillas c:

mmur07 committed:
fixes del uso de rand(), sprite del jugador e includes

fixes menores adicionales

mmur07 commented:
Ya están resueltos los cambios pedidos.

EstebenR commented on src/SourceFiles/ObjectFactory.h:
```diff
@@ -1,18 +1,24 @@
+class InputHandler;
 class Weapon;
+class Hands;
+class Coin;
```
Si las clases no se usan en el .h al igual que no hay que incluirlas no hay que poner class. En este caso solo haria falta EntityManager que es lo unico a lo que hace mencion el ObjectFactory.h

EstebenR commented on src/SourceFiles/CapitalismGameMode.cpp:
```diff
@@ -0,0 +1,38 @@
+}
+
+void CapitalismGameMode::update() {
+	timeSinceStart += CONST(double, "MS_PER_FRAME");
+	if (timeSinceStart >= CONST(double, "TIME_TO_END") && !roundFinished_) {
```
Las const pedirlas en el init y guardarlo, que es un proceso algo costoso estar pidiendolo cada ciclo ya que tiene que ir al archivo y buscarlo.

EstebenR reviewed (changes requested):
Lo de las const revisadlo en otros archivos que sale en algún otro. Mañana sigo que tengo sueño, besitos

mmur07 committed:
requests arreglados

arreglado el uso de constantes y las inclusiones de archivos

EstebenR commented on src/SourceFiles/WiFightGameMode.cpp:
```diff
@@ -4,8 +4,10 @@
+	playerProgress_.reserve(nPlayers_);
+	for (int k = 0; k < nPlayers_; k++) {
+		players_.push_back(PlayerFactory::createPlayerWithHealth(game->getEntityManager(), game->getPhysicsWorld(), k,
+			Resources::Tinky, tilemap_->getPlayerSpawnPoint(k).x, tilemap_->getPlayerSpawnPoint(k).y, 3));
```
Ste men sigue siendo un tinky

EstebenR commented on src/SourceFiles/Wallet.cpp:
```diff
@@ -0,0 +1,23 @@
+	cout << "HE SOLTADO " << coinsToDrop << endl;
+	for (int k = 0; k < coinsToDrop; k++) {
+		double a = (rand() % 360 * CONST(double, "PI")) / 180;
+		double x = (collPlayer_->getH(0) + CONST(double, "COIN_BASE_SIZE"))* cos(a);
+		double y = (collPlayer_->getH(0) + CONST(double, "COIN_BASE_SIZE")) * sin(a);
```
Pedida de constantes

EstebenR commented on src/SourceFiles/CoinsPool.h:
```diff
@@ -0,0 +1,22 @@
+#pragma once
+#include "ObjectPool.h"
+#include "Entity.h"
+#include "SDL_Game.h"
+#include "ObjectFactory.h"
+#include "EntityManager.h"
+#include "b2_world.h"
+#include "Viewer.h"
+#include "Coin.h"
+#include "b2_polygon_shape.h"
```
include de un monton de headers, antes de simplemente cortar y pegar mirar si son todos necesarios

EstebenR commented on src/SourceFiles/StocksGameMode.cpp:
```diff
@@ -12,9 +14,13 @@ StocksGameMode::~StocksGameMode()
+			Resources::Body, tilemap_->getPlayerSpawnPoint(i).x, tilemap_->getPlayerSpawnPoint(i).y, 3));
 		playerStocks_.push_back(maxStocks_); //Initializes maxStocks vector with 3 on all positions.
 	}
+	for(int i = 0;i<players_.size();i++){
+	}
```
Sobra este bucle?

EstebenR commented on src/SourceFiles/Coin.cpp:
```diff
@@ -0,0 +1,28 @@
+void Coin::update()
+{
+	if (justDroppedByPlayer_ != -1) {
+		timeSinceDropped_ += CONST(double, "MS_PER_FRAME");
+		if (timeSinceDropped_ >= CONST(int, "COINS_COOLDOWN_WHEN_DROPPED")) {
```
consts

EstebenR reviewed (changes requested):
Los personajes salen achatados, revisad el tamaño cuando los creáis

mmur07 committed:
problemas con constantes y sprites finalizado y espero que el ping-pong para los includes tambien

title.

mmur07 committed:
arreglo constante pi

Auraum commented on src/SourceFiles/WiFightGameMode.cpp:
```diff
@@ -4,8 +4,10 @@
 void WiFightGameMode::init(PlayState* game)
```
Por qué no creáis el router con el ObjectFactory?

Auraum reviewed (changes requested):
Da error al cerrar el juego con la X y no me convence mucho que cada jugador colisione con las monedas que desprende, debería poder recuperarlas. Pero me gusta lo de que los players los haga el gameMode.

Pj6595 committed:
Algunos fixes

Pj6595 committed:
Arreglado error de salida de juego

mmur07 committed:
Arreglo de las monedas y de las armas a meele

Las monedas funcionan de acuerdo a lo hablado en el scrum pasado.
Ahora las armas a meele te quitan porcentaje de monedas.

mmur07 committed:
Arreglado el agarre a objetos

La comprobación de agarrarse a objetos asumía que en la colisión que se producía, el fixA era el player y el fixB la pared/objeto agarrable. Al haber cambiado las capas de colisiones (creo), se han invertido estos 2 valores. Por tanto, hemos mejorado la comprobación del agarre para que tenga en cuenta ambos casos.

*a

JuanmaSalva committed:
Puede q los spawns por tiled funcuinen

Es el push de schrodinger

Pj6595 committed:
Modo Capitalismo 100% definitivo 1 link real

Pj6595 commented:
Mañana seguimos con la UI que tenemos sueño. Besitos c:

EstebenR reviewed (approved):
Molto bene, me gustaria que las roombas se dibujaran debajo de los jugadores

Auraum reviewed (approved):
muy chulo

Pj6595 committed:
Los roombas se dibujan por debajo de los jugadores

AdriAlvarez00 reviewed (approved):
No he coseguido que se rompa
Está casi tan desbalanceado como el capitalismo de verdad <3
